### ISSUE TITLE
.is_in() inconsistent behavior with input list.

### ISSUE THREAD

**Issue by jesusestevez** [pola-rs/polars#9703]
### Polars version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Issue description

We have noticed that .is_in() provides wrong results when inputting a list that does not contain a column in the dataframe.

### Reproducible example

```python
import polars as pl

df = pl.DataFrame(
    {"A": [[1.0, 2.0, 3.0], [1.0, 9.0], [9.0, 3.0]],
     "C": [1.0, 2.0, 3.0],
     "B": [8.0, 8.0, 3.0]}
)
print(df.with_columns([pl.col("C").is_in(pl.concat_list([pl.col("B"), 8.0])).alias("contains")]))
print(df.with_columns([pl.col("C").is_in(pl.concat_list([1.0, 8.0])).alias("contains")]))

shape: (3, 4)
┌─────────────────┬─────┬─────┬──────────┐
│ A               ┆ C   ┆ B   ┆ contains │
│ ---             ┆ --- ┆ --- ┆ ---      │
│ list[f64]       ┆ f64 ┆ f64 ┆ bool     │
╞═════════════════╪═════╪═════╪══════════╡
│ [1.0, 2.0, 3.0] ┆ 1.0 ┆ 8.0 ┆ false    │
│ [1.0, 9.0]      ┆ 2.0 ┆ 8.0 ┆ false    │
│ [9.0, 3.0]      ┆ 3.0 ┆ 3.0 ┆ true     │
└─────────────────┴─────┴─────┴──────────┘
shape: (3, 4)
┌─────────────────┬─────┬─────┬──────────┐
│ A               ┆ C   ┆ B   ┆ contains │
│ ---             ┆ --- ┆ --- ┆ ---      │
│ list[f64]       ┆ f64 ┆ f64 ┆ bool     │
╞═════════════════╪═════╪═════╪══════════╡
│ [1.0, 2.0, 3.0] ┆ 1.0 ┆ 8.0 ┆ true     │
│ [1.0, 9.0]      ┆ 2.0 ┆ 8.0 ┆ true     │
│ [9.0, 3.0]      ┆ 3.0 ┆ 3.0 ┆ true     │
└─────────────────┴─────┴─────┴──────────┘
```

### Expected behavior

We would expect the expression to behave in such a way that the rows in column C are evaluated rowwise vis-à-vis the list, following the approach of the first table printed. However, currently it seems that only the first element of the list is being evaluated, as the second example shows.

We consider this may be a potential bug in the expression, given that `df.with_columns(pl.concat_list([1.0, 8.0]))` would add a column with the list being in every row. 

### Installed versions

<details>

```
--------Version info---------
Polars:      0.18.2
Index type:  UInt32
Platform:    Windows-10-10.0.19044-SP0
Python:      3.11.3 | packaged by conda-forge | (main, Apr  6 2023, 08:50:54) [MSC v.1934 64 bit (AMD64)]

----Optional dependencies----
numpy:       1.24.3
pandas:      2.0.1
pyarrow:     12.0.0
connectorx:  <not installed>
deltalake:   <not installed>
fsspec:      2021.11.1
matplotlib:  3.7.1
xlsx2csv:    <not installed>
xlsxwriter:  3.1.2
```

</details>

**Comment by cmdlineluser** [comment]
Using `.select()` may help debug:
```python
df = pl.DataFrame({
     "A": [1.0, 2.0, 3.0],
     "B": [8.0, 8.0, 3.0]
})

df.select(pl.col("A").is_in([1.0]))
# shape: (3, 1)
# ┌───────┐
# │ A     │
# │ ---   │
# │ bool  │
# ╞═══════╡
# │ true  │
# │ false │
# │ false │
# └───────┘
```
The `.concat_list` is causing it to be a column-based comparison:
```python
df.select(pl.col("A").is_in(pl.concat_list([1.0])))

# shape: (1, 1)
# ┌──────┐
# │ A    │
# │ ---  │
# │ bool │
# ╞══════╡
# │ true │
# └──────┘
```

When using `.with_columns()` this single value is then broadcasted to fit the length of the dataframe:

```python
df.with_columns(pl.col("A").is_in(pl.concat_list([1.0])))

# shape: (3, 2)
# ┌──────┬─────┐
# │ A    ┆ B   │
# │ ---  ┆ --- │
# │ bool ┆ f64 │
# ╞══════╪═════╡
# │ true ┆ 8.0 │
# │ true ┆ 8.0 │
# │ true ┆ 3.0 │
# └──────┴─────┘
```

**Comment by jesusestevez** [comment]
Thank you very much for the suggestion. I can see, however, that we could use the output of concat_list() as an input to a subsequent expression using is_in(). For example:

print(df.with_columns(pl.concat_list([pl.col("B"), 8.0]).alias("helper")).with_columns([pl.col("C").is_in(pl.col("helper")).alias("contains")]))

shape: (3, 5)
┌─────────────────┬─────┬─────┬────────────┬──────────┐
│ A               ┆ C   ┆ B   ┆ helper     ┆ contains │
│ ---             ┆ --- ┆ --- ┆ ---        ┆ ---      │
│ list[f64]       ┆ f64 ┆ f64 ┆ list[f64]  ┆ bool     │
╞═════════════════╪═════╪═════╪════════════╪══════════╡
│ [1.0, 2.0, 3.0] ┆ 1.0 ┆ 8.0 ┆ [8.0, 8.0] ┆ false    │
│ [1.0, 9.0]      ┆ 2.0 ┆ 8.0 ┆ [8.0, 8.0] ┆ false    │
│ [9.0, 3.0]      ┆ 3.0 ┆ 3.0 ┆ [3.0, 8.0] ┆ true     │
└─────────────────┴─────┴─────┴────────────┴──────────┘

This means that concat_list() generates indeed a column with the single value (in this case a list) broadcasted to fit the length of the DataFrame, where polars flags properly that "B" is a column, while the float *.o is just concatenated in the dataframe.

I am still missing the link between the column based comparison, which makes sense in the context of the function, and the behavior of the polars expression above, as I would expect it to create a single list of items and return a true for all the rows in the column contains. 
In other words, I would expect the column helper to return, as you said, a single list of items such that: [8.0, 8.0, 3.0, 8.0] which would then be broadcasted

**Comment by avimallu** [comment]
I find [.inspect](https://pola-rs.github.io/polars/py-polars/html/reference/expressions/api/polars.Expr.inspect.html) to be a handy tool to evaluate what's going wrong (or right). Let's take your example DataFrame:

```py
import polars as pl

df = pl.DataFrame(
    {"A": [[1.0, 2.0, 3.0], [1.0, 9.0], [9.0, 3.0]],
     "C": [1.0, 2.0, 3.0],
     "B": [8.0, 8.0, 3.0]}
)
```

See what happens in second run of `pl.concat_list`:

```py
(
    df
    .with_columns(
        pl.col("C").is_in(
            pl.concat_list([1.0, 8.0])
            .inspect()
        ).alias("contains"))
)
```

In addition to the output, it gives from `.inspect` (I've reformatted the code to make it easier to follow):

```py
shape: (1,)
Series: 'literal' [list[f64]]
[
	[1.0, 8.0]
]
```

which means that @cmdlineluser's hypothesis is right - a column based comparison happens. Why? `pl.concat_list` returns an expression, which evaluates in the *context* of the DataFrame to become a Polars `List[f64]` of length 1. Being length 1, it gets broadcasted to the length of the DataFrame.

How do I deduce that? We've established from `inspect` that a `Series` object of length 1 is provided to `is_in` to operate on. Let's manually give a `Series` object of length 2:

```py
(
    df
    .with_columns(
        pl.col("C").is_in(
            pl.Series("a", [[1.0, 8.0], [1.0, 8.0]])
        ).alias("contains"))
)
```

This will fail, and give the error: `ShapeError: unable to add a column of length 2 to a dataframe of height 3`

What happens in the example you showed in your response? `pl.concat` now still returns an expression, still returns a `List[f64]`, but with length equal to the length of column B. You materialize it, so it gets added as the `helper` column. You can see what happens without the materialization below:

```py
(
    df
    .with_columns(
        pl.col("C").is_in(
            pl.concat_list(
                pl.col("B"), 8.0)
            .inspect())
        .alias("contains"))
)
```

Gives the `.inspect` output:

```py
shape: (3,)
Series: 'B' [list[f64]]
[
	[8.0, 8.0]
	[8.0, 8.0]
	[3.0, 8.0]
]
```

> I would expect the column helper to return, as you said, a single list of items such that: [8.0, 8.0, 3.0, 8.0] which would then be broadcasted

When would this kind of an operation happen? When you `explode` the list, create a `Series` of `Float64` values, which don't have to fit to a DataFrame, so becomes the `Series` argument type that `is_in` expects. See below:

```py
(
    df
    .with_columns(
        pl.col("C").is_in(
            pl.concat_list(
                pl.col("B"), 8.0)
            .list.explode()
            .inspect())
        .alias("contains"))
)
```

Gives the`.inspect` output:

```py
shape: (6,)
Series: 'B' [f64]
[
	8.0
	8.0
	8.0
	8.0
	3.0
	8.0
]
```

**Comment by jesusestevez** [comment]
Thank you very much for the detailed explanation. Now we understand the process better, and consider that we can close the issue as it is not a bug.
